### PR TITLE
`this` context within handlers

### PR DESCRIPTION
**Pull request by jsumners**
While working on `fastify-caching` I discovered some oddness with the decorated instance. Specifically, when registering a plugin that uses `fastify-plugin` the decorated properties are accessible via the `instance` reference but not via `this` within handlers.

This PR is to get feedback on solving this issue.

### CODE REVIEW

**Comment by allevo** [comment]
Truly the `this` inside the handler should be the `context` instance (see [here](https://github.com/fastify/fastify/blob/master/lib/handleRequest.js#L120))

Do you want to move `this` to `fastify` instance?

**Comment by jsumners** [comment]
Benchmarks

**master:**
```
% npm_config_usingfile=simple.js npm run benchmark 
> fastify@0.32.0 benchmark /Users/jsumners/Projects/oss/fastify
> node ./examples/$npm_config_usingfile & pid=$! && autocannon -c 100 -d 5 -p 10 localhost:3000/ && kill -9 $pid

Running 5s test @ http://localhost:3000/
100 connections with 10 pipelining factor

running [                    ] 0%
server listening on 3000
Running 5s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 4.89    9.74    314
Req/Sec      19245.6 2880.24 20991
Bytes/Sec    2.86 MB 433 kB  3.15 MB

96k requests in 5s, 14.3 MB read
15 errors (0 timeouts)
```

**branch:**
```
% npm_config_usingfile=simple.js npm run benchmark

> fastify@0.32.0 benchmark /Users/jsumners/Projects/oss/fastify
> node ./examples/$npm_config_usingfile & pid=$! && autocannon -c 100 -d 5 -p 10 localhost:3000/ && kill -9 $pid

Running 5s test @ http://localhost:3000/
100 connections with 10 pipelining factor

running [                    ] 0%
server listening on 3000
Running 5s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 5.1     10.24   328
Req/Sec      18464.8 2896.85 20527
Bytes/Sec    2.76 MB 449 kB  3.15 MB

92k requests in 5s, 13.8 MB read
16 errors (0 timeouts)
```

Hmm, there has to be a better way.

File: fastify.js [hunk 1]
```diff
@@ -408,6 +410,10 @@ function build (options) {
         opts.middie || _fastify._middie
       )
 
+      for (var decoration of _fastify._decorations) {
```
Comment (StarpTech) [inline fastify.js]: with that its not possible to define decorators in routes right?

**Comment by jsumners** [comment]
@mcollina would you mind pushing a commit to this branch which does that?

**Comment by StarpTech** [comment]
@jsumners here 
use `handler.bind(self)`
https://github.com/fastify/fastify/blob/master/fastify.js#L348

**Comment by jsumners** [comment]
New branch benchmarks:

```
server listening on 3000
Running 5s test @ http://localhost:3000/
100 connections with 10 pipelining factor

Stat         Avg     Stdev   Max
Latency (ms) 4.75    9.58    318
Req/Sec      19764   2945.14 21503
Bytes/Sec    2.96 MB 449 kB  3.28 MB

99k requests in 5s, 14.7 MB read
11 errors (0 timeouts)
```

**Comment by StarpTech** [comment]
LGTM

**Comment by jsumners** [comment]
Excellent. Will need a new release before I can continue with the caching plugin.

**Comment by delvedor** [comment]
Landed in [`v0.33.0`](https://github.com/fastify/fastify/releases/tag/v0.33.0) :)

**Comment by jsumners** [comment]
👍

**Pull request merged**